### ISSUE TITLE
Improvements to error messaging 

### ISSUE THREAD

**Issue by stephanieelliott** [Expensify/Expensify.cash#1091]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!

___

View the job in Upwork here. 
**Platform - version:**  ALL 

----
**Action Performed (reproducible steps):**
Summary: Currently, when a user encounters a sign-in error on expensify.cash, we display the raw API message to the user. We'd like to replace these responses so that when an error occurs, we display a user-friendly message instead. 

The error codes and associated user-friendly messages are included in the `Messaging` section of this issue. 

**Steps to reproduce an example of the issue:**
1. Go to expensify.cash
2. Sign in using an incorrect email address and password combination (i.e. a pair of credentials not associated with an account)
3. See error code and generic error message appear (in red text). 
     - Example: `401 No user with that partner/user/password`

**Desired Result:**
Show user-friendly error messages. Example: 401 error should show: `Incorrect password`

**Actual Result**
Generic  `401 No user with that partner/user/password` appears 

**Notes/Photos/Videos:** 
![image](https://user-images.githubusercontent.com/31225194/103105158-e53bb200-45f9-11eb-85b7-1746746e4139.png)
 
----

**Messaging:**

| Error Code  | Message to display |
| ------------- | ------------- |
| 401  | Incorrect login or password. Please try again. |
| 402  | You have 2FA enabled on this account. Please sign in using your email or phone number. |
| 403 | Invalid login or password. Please try again or reset your password. | 
| 404 | We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes. | 
| 405 |  You do not have access to this application. Please add your GitHub username for access. |

**Comment by sidhiartha** [comment]
let me try to help you fixing this, I have submit the proposal on upwork under Sidhi,
let me rephrase the date expectation, it is within 1 day after I can run the project successfully.

**Comment by sidhiartha** [comment]
let me try to help you fixing this @stephanieelliott , 
I have submit the proposal on upwork under Sidhi,
let me rephrase the date expectation, 
it is within 1 day after I can run the project successfully.

**Comment by anthony-hull** [comment]
Telling the user that their password is wrong is giving away important information.
It tells the requestor that the username exists but that the password is wrong. Giving a non-specific error that the username or password is wrong is better security practice. but a worse UX.

Either the UX or Security suffers here. Although if the API is giving those codes I'm guessing this choice has already been made?

**Comment by billiramirez** [comment]
I agree with @anthony-hull , it's better having a more generic message, for instance: "There was an error signing to the App"

**Comment by stephanieelliott** [comment]
Good points, thanks @billiramirez and @anthony-hull. I updated the `401` messaging to be more ambiguous as to which credential is failing.

**Comment by ajay426** [comment]
I am working on this.

**Comment by ajay426** [comment]
Hey @stephanieelliott
I am also getting `413 Account locked.` after so many wrong password attempts. Please let me know the error message for it as well.

**Comment by ajay426** [comment]
As a proposal of the solution to the problem, I am thinking of making a common function to extract the error message from the API error response. As I have observed from the error message format, it includes the error code at the start of the error message, I can extract the error code from the start of the error message. Then after getting the error code, I have made a `switch-case` statement to return the respective error message based on the error code returned in the API error response. If in future, we would have to update the error message, or include another status code, then we only have to look into this function and update it as per the requirement. It may look like this:

```

function getErrorMessage(error = '') {
  const errorCode = error.split(' ')[0];

  switch (errorCode) {
    case '401':
      return 'Incorrect login or password. Please try again.'
    case '402':
      return 'You have 2FA enabled on this account. Please sign in using your email or phone number.'
    case '403':
      return 'Invalid login or password. Please try again or reset your password.'
    case '404':
      return 'We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes.'
    case '405':
      return 'You do not have access to this application. Please add your GitHub username for access.'
    case '413':
      return 'Your account has been locked after so many unsuccessfull attempts. Please try again after 1 hour.';

    default:
      return error;
  }
}
```
If the error message does not include the error code at the start, then the function will return the error message as it is from the API response.

And the usage of the function could be like:

```
const error = '413 Account locked.';
console.log(getErrorMessage(error));
```

As much I have reviewed the code, the location of the usage of this function could be in the file `src/libs/actions/Session.js` at `line number 159` like:

```
Onyx.merge(ONYXKEYS.SESSION, {error: getErrorMessage(error.message)});
```

Please let me know about your views.

**Comment by devkashan1** [comment]
Hey everyone! The solution I have planned will make 2 different files for this.

1. The First file will contain all the static messages in JSON format along with Type, ErrorCode, and Message.
2. The Second file will have a couple of functions where we can setup messages and types to fetch and call them into the components in a single line and single method with parsing error code and their types.

It will make this whole functionality reusable. Please let me know if you have any questions.

**Comment by devkashan1** [comment]
I'm starting to work it.

**Comment by devkashan1** [comment]
The solution I have planned will make 2 different files for this.

1. The First file will contain all the static messages in JSON format along with Type, ErrorCode, and Message.
2. The Second file will have a couple of functions where we can setup messages and types to fetch and call them into the components in a single line and single method with parsing error code and their types.

It will make this whole functionality reusable. Please let me know if you have any other questions.

**Comment by devkashan1** [comment]
The details of the implementation for this problem that I have fixed in the project as I briefed above:

I have added a new file where we can manage any error messages at this place `src\libs\ErrorMessage\messages.js`.

```
const messages = [
  {
    type: 'Login',
    errorCode: 401,
    message: 'Incorrect login or password. Please try again.',
  },
  {
    type: 'Login',
    errorCode: 402,
    message:
      'You have 2FA enabled on this account. Please sign in using your email or phone number.',
  },
  {
    type: 'Login',
    errorCode: 403,
    message:
      'Invalid login or password. Please try again or reset your password.',
  },
  {
    type: 'Login',
    errorCode: 404,
    message:
      'We were unable to change your password. This is likely due to an expired password reset link in an old password reset email. We have emailed you a new link so you can try again. Check your Inbox and your Spam folder; it should arrive in just a few minutes.',
  },
  {
    type: 'Login',
    errorCode: 405,
    message:
      'You do not have access to this application. Please add your GitHub username for access.',
  },
  {
    type: 'Login',
    errorCode: 413,
    message: 'Your account has been locked.',
  },
];
export default messages;
```

I have also added another file where I created a function to fetch the appropriate message via type and error code. The path is `src\libs\ErrorMessage\index.js.` 
```
import _ from 'lodash';
import messages from './messages';

/**
 * It's generic function to print any messages.
 *
 * @param {String} type
 * @param {String} error
 * @returns {String}
 */
function getErrorMessage(type, error) {
  const code = error.split(' ')[0];
  if (!_.isEmpty(code)) {
    const message = _.filter(messages, {type: type, errorCode: +code.trim()});
    if (!_.isEmpty(message)) {
      return message[0].message;
    }
  }
  return error;
}
export {getErrorMessage};

```

After these, I have modified a file `src\libs\actions\Session.js` where we call the `getErrorMessage` function.

Line# 11:  `import {getErrorMessage} from '../ErrorMessage';`
Line# 162:  `Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage('Login',error.message)});`
Line# 169:  ` Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage('Login',error.message), loading: false});`

That's all.

Please let me know if you have any questions. I'll create a PR and post a message in slack shortly.

**Comment by devkashan1** [comment]
Hey @stephanieelliott  I saw there is another error message with the code like `413 Account locked.` For now, I have changed the message to `Your account has been locked.`  If you want to change this message, please let me know. It's customizable. Thank you.

**Comment by ajay426** [comment]
Hello @stephanieelliott!
Should I also proceed with creating a PR?
I am still waiting for my proposed solution to be reviewed as per mentioned in the procedure to submit the solution.

**Comment by yuwenmemon** [comment]
@ajay426 your proposal looks good, and yes please use the `jsonCode` in the response from the API.

**Comment by yuwenmemon** [comment]
@devkashan1 What the purpose of the `type` parameter in your error messages object? Currently, looks like they are all the same?

`messages` seems like a bit of a generic name, I think this might be better as `errorMessages`? 

Overall, the scope of this issue is simply to improve the error messaging upon login. It appears your solution is attempting to establish a generalized system for which all error messages are handled, and I think that's probably more than we need.

**Comment by devkashan1** [comment]
@yuwenmemon  The `type` parameter is used to identify the screen on which error messages need to be shown. In this way, we can use it on any screen.

It's not a big deal. Do you want to change it?

Yes, You're right. We can use it anywhere in the project, and It's the best approach for handling error messages.

**Comment by ajay426** [comment]
Can you please let me know how can I reproduce the use case for the error codes `402`, `403`, `404` and `405`?
What are the steps to reproduce

**Comment by yuwenmemon** [comment]
@devkashan1 I would advise changing. It's probably overkill for what we need right now.

**Comment by yuwenmemon** [comment]
@ajay426 some of them may be harder to reproduce than others because of restrictions we have on the client. I would say establishing that the code works for one or two cases should be enough. I think I know how you can reproduce `405`...

`405`: Try logging in with an account that does not have access to the application but is an expensify account. You can create a "dummy" account on expensify.com (simply enter an email into the sign in, then log out right away. This will create an invalidated "dummy" account) and then try logging into the expensify.cash site.

**Comment by devkashan1** [comment]
@yuwenmemon I have made and pushed the required changes. Please review.

**Comment by yuwenmemon** [comment]
> @yuwenmemon I have made and pushed the required changes. Please review.

@devkashan1 apologies, but where?

**Comment by devkashan1** [comment]
> > @yuwenmemon I have made and pushed the required changes. Please review.
> 
> @devkashan1 apologies, but where?

Yeah, Here is the PR https://github.com/Expensify/Expensify.cash/pull/1166 that have pushed the changes. Should I share the details that I made?

**Comment by yuwenmemon** [comment]
Thanks, @devkashan1. If you would, please include the issue that this PR fixes in the PR description.

**Comment by devkashan1** [comment]
> Thanks, @devkashan1. If you would, please include the issue that this PR fixes in the PR description.

Sure things. I'll do now.

**Comment by devkashan1** [comment]
@yuwenmemon I have added the description in the PR. please check https://github.com/Expensify/Expensify.cash/pull/1166#issuecomment-759067950

**Comment by yuwenmemon** [comment]
Contributors were both paid in Upwork. The contract has been ended in Upwork.
